src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Services/BookmarkResolver.cs

```diff
@@ -0,0 +1,548 @@
+    {
+        try
+        {
+            // This goes greedy from the longest head down to shortest; exactly opposite of what
+            // CreateProcess rules are for the first token. But here we operate with a slightly different goal.
+            var (greedyHead, greedyTail) = GreedyFind(head, containsPlaceholders, placeholderParser);
+
+            // put tails back together:
+            return (Head: greedyHead, string.Join(" ", greedyTail, tail).Trim());
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError("Failed to find best path", ex);
+            throw;
+        }
+    }
+
+    private static (string? Head, string? Tail) GreedyFind(string input, bool containsPlaceholders, IPlaceholderParser placeholderParser)
+    {
+        // Be greedy: try to find the longest existing path prefix
+        for (var i = input.Length; i >= 0; i--)
+        {
+            if (i < input.Length && !char.IsWhiteSpace(input[i]))
+            {
+                continue;
+            }
+
+            var candidate = input.AsSpan(0, i).TrimEnd().ToString();
+            if (candidate.Length == 0)
+            {
+                continue;
+            }
+
+            // If we have placeholders, check if this candidate would contain a non-path placeholder
+            if (containsPlaceholders && ContainsNonPathPlaceholder(candidate, placeholderParser))
+            {
+                continue; // Skip this candidate, try a shorter one
+            }
+
+            try
+            {
+                if (CommandLineHelper.ExpandPathToPhysicalFile(candidate, true, out var full))
+                {
+                    var tail = i < input.Length ? input[i..].TrimStart() : string.Empty;
+                    return (full, tail);
+                }
+            }
+            catch
+            {
+                // Ignore malformed paths; keep scanning
+            }
+        }
+
+        return (null, null);
+    }
+
+    private static bool ContainsNonPathPlaceholder(string candidate, IPlaceholderParser placeholderParser)
+    {
+        placeholderParser.ParsePlaceholders(candidate, out _, out var placeholders);
+        foreach (var match in placeholders)
+        {
+            var placeholderContext = GetPlaceholderContextInFileSystemPath(candidate, match.Index);
+
+            // If placeholder appears after what looks like a command-line flag/option
+            if (placeholderContext.IsAfterFlag)
+            {
+                return true;
+            }
+
+            // If placeholder doesn't look like a typical path component
+            if (!placeholderContext.LooksLikePathComponent)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static PlaceholderContext GetPlaceholderContextInFileSystemPath(string input, int placeholderIndex)
```

Comment:
(imo this could use a comment, because this is basically just "let's make an educated guess")